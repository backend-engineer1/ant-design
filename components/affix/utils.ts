import addEventListener from 'rc-util/lib/Dom/addEventListener';
import getScroll from '../_util/getScroll';
import Affix from './';

// ======================== Observer ========================
const TRIGGER_EVENTS = [
  'resize',
  'scroll',
  'touchstart',
  'touchmove',
  'touchend',
  'pageshow',
  'load',
];

interface ObserverEntity {
  target: HTMLElement | Window;
  affixList: Affix[];
  eventHandlers: { [eventName: string]: any };
}

let observerEntities: ObserverEntity[] = [];

export function addObserveTarget(target: HTMLElement | Window | null, affix: Affix): void {
  console.log('???', target);
  if (!target) return;

  let entity: ObserverEntity | undefined = observerEntities.find(item => item.target === target);

  if (entity) {
    entity.affixList.push(affix);
  } else {
    entity = {
      target,
      affixList: [affix],
      eventHandlers: {},
    };
    observerEntities.push(entity);

    // Add listener
    TRIGGER_EVENTS.forEach(eventName => {
      entity!.eventHandlers[eventName] = addEventListener(target, eventName, (event: Event) => {
        entity!.affixList.forEach(affix => {
          affix.updatePosition(event);
        });
      });
    });
  }
}

export function removeObserveTarget(affix: Affix): void {
  const observerEntity = observerEntities.find(oriObserverEntity => {
    const hasAffix = oriObserverEntity.affixList.some(item => item === affix);
    if (hasAffix) {
      oriObserverEntity.affixList = oriObserverEntity.affixList.filter(item => item !== affix);
    }
    return hasAffix;
  });

  if (observerEntity && observerEntity.affixList.length === 0) {
    observerEntities = observerEntities.filter(item => item !== observerEntity);

    // Remove listener
    TRIGGER_EVENTS.forEach(eventName => {
      const handler = observerEntity.eventHandlers[eventName];
      if (handler && handler.remove) {
        handler.remove();
      }
    });
  }
}

export function getTargetRect(target: HTMLElement | Window | null): ClientRect {
  return target !== window
    ? (target as HTMLElement).getBoundingClientRect()
    : ({ top: 0, bottom: window.innerHeight } as ClientRect);
}

export function getOffset(element: HTMLElement, target: HTMLElement | Window | null) {
  const elemRect = element.getBoundingClientRect();
  const targetRect = getTargetRect(target);

  const scrollTop = getScroll(target, true);
  const scrollLeft = getScroll(target, false);

  const docElem = window.document.body;
  const clientTop = docElem.clientTop || 0;
  const clientLeft = docElem.clientLeft || 0;

  return {
    top: elemRect.top - targetRect.top + scrollTop - clientTop,
    left: elemRect.left - targetRect.left + scrollLeft - clientLeft,
    width: elemRect.width,
    height: elemRect.height,
  };
}
