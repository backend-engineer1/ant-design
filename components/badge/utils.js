import React from 'react';

export function toArrayChildren(children) {
  const ret = [];
  React.Children.forEach(children, (c) => {
    ret.push(c);
  });
  return ret;
}

export function getPartNumber(num) {
  if (!num) {
    return null;
  }
  const countStr = num.toString();
  const obj = {};
  for (let i = 0; i < countStr.length; i++) {
    obj[countStr.length - 1 - i] = Number(countStr[i]);
  }
  return obj;
}

export function getTranslateY(differ, data, _data, height, i) {
  let translateY = 0;
  if (!differ) {
    //不想插入40个，改变要滚到的距离；
    if (_data[i + '_add']) {
      return -(data[i] + 20) * height;
    }
    if (_data[i + '_rem']) {
      return -data[i] * height;
    }
    return false;
  }
  //判断相差的位数来驱动数字；差个位滚动一格，差10的部分及以上个位滚动一周；
  const countToString = Math.abs(differ).toString();
  const countLength = countToString.length;
  const on = _data[i];
  const to = data[i];
  if (differ > 0) {
    if (countLength - 1 > i) {
      translateY = -(to - (to - on)) * height;
      //on大于to且differ大于10，如9->0,需要设计滚动到的位置＋10;
      if (on > to) {
        data[i + '_add'] = true;
      }
    } else if (countLength - 1 < i && to === on) {
      //参数不变动；
      translateY = null;
    } else if (typeof on === 'undefined') {
      //新增加入时设为0；
      translateY = -10 * height;
    } else {
      //如果开始大于到达，到达（to）加10；
      const _to = on > to ? to + 10 : to;
      translateY = -(to - (_to - on) + 10) * height;
    }
  } else {
    if (countLength - 1 > i) {
      translateY = -(to + (on - to) + 20) * height;
      //同上，differ大于10时，且to在于on
      if (to > on) {
        data[i + '_rem'] = true;
      }
    } else if (countLength - 1 < i && to === on) {
      translateY = null;
    } else {
      const _on = on < to ? on + 10 : on;
      translateY = -(to + (_on - to) + 10) * height;
    }
  }
  return translateY;
}
