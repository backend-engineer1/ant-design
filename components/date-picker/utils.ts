import * as moment from 'moment';

export function formatDate(
  value: moment.Moment | undefined | null,
  format: string | string[],
): string {
  if (!value) {
    return '';
  }
  if (Array.isArray(format)) {
    format = format[0];
  }
  return value.format(format);
}
