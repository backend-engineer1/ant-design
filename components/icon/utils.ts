// These props make sure that the SVG behaviours like general text.
// Reference: https://blog.prototypr.io/align-svg-icons-to-text-and-say-goodbye-to-font-icons-d44b3d7b26b4
export const svgBaseProps = {
  width: '1em',
  height: '1em',
  fill: 'currentColor',
  ['aria-hidden']: 'true',
};
