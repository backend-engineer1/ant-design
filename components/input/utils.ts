import type { InputProps } from './Input';

// eslint-disable-next-line import/prefer-default-export
export function hasPrefixSuffix(props: InputProps) {
  return !!(props.prefix || props.suffix || props.allowClear);
}
