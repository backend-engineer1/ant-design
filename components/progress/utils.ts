import { presetPrimaryColors } from '@ant-design/colors';
import type { CircleProps } from './Circle';
import type { ProgressProps } from './progress';
import warning from '../_util/warning';

export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    warning(
      false,
      'Progress',
      '`success.progress` is deprecated. Please use `success.percent` instead.',
    );
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export const getPercentage = ({ percent, success, successPercent }: ProgressProps) => {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
};

export const getStrokeColor = ({ success = {}, strokeColor }: Partial<CircleProps>): (
  | string
  | Record<PropertyKey, string>
)[] => {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
};

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
    strokeWidth?: number;
  },
): [number, number] => {
  let width: number = -1;
  let height: number = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    const strokeWidth = extra!.strokeWidth!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = strokeWidth ?? 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = size;
    }
    width *= steps;
  } else if (type === 'line') {
    const strokeWidth = extra?.strokeWidth;
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = strokeWidth || (size === 'small' ? 6 : 8);
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = size;
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      if (process.env.NODE_ENV !== 'production') {
        warning(
          false,
          'Progress',
          'Type "circle" and "dashbord" do not accept array as `size`, please use number or preset size instead.',
        );
      }

      width = size[0] ?? size[1] ?? 120;
      height = size[0] ?? size[1] ?? 120;
    }
  }
  return [width, height];
};
