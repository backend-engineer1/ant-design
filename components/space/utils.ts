import type { SpaceSize } from '.';
import type { SizeType } from '../config-provider/SizeContext';

export function isPresetSize(size: SpaceSize): size is SizeType {
  return ['small', 'middle', 'large'].includes(size as string);
}

function isValidGapNumber(size: SpaceSize): size is number {
  if (!size) {
    // The case of size = 0 is deliberately excluded here, because the default value of the gap attribute in CSS is 0, so if the user passes 0 in, we can directly ignore it.
    return false;
  }
  return typeof size === 'number' && !Number.isNaN(size);
}

export const getRealSize = (
  sizeMap: Record<NonNullable<SpaceSize>, number>,
  value: SpaceSize,
): number => {
  if (isPresetSize(value)) {
    return sizeMap[value!];
  }
  if (isValidGapNumber(value)) {
    return value;
  }
  return 0;
};
