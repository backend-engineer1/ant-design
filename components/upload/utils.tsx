import { RcFile, UploadFile } from './interface';

export function T() {
  return true;
}

type WrapFile = RcFile | UploadFile;

/**
 * Wrap file with Proxy to provides more info. Will fallback to object if Proxy not support.
 *
 * Origin comment: Fix IE file.status problem via coping a new Object
 */
export function wrapFile(file: WrapFile): UploadFile {
  const filledProps = {
    lastModified: file.lastModified,
    lastModifiedDate: file.lastModifiedDate,
    name: file.name,
    size: file.size,
    type: file.type,
    uid: file.uid,
    percent: 0,
    originFileObj: file,
  };

  if (typeof Proxy !== 'undefined') {
    const data = new Map<string | symbol, any>(Object.entries(filledProps));

    const getValue = (key: string | symbol) => {
      if (data.has(key)) {
        return data.get(key);
      }
      return (file as any)[key];
    };

    return new Proxy(file, {
      get(_, key) {
        return getValue(key);
      },
      set(_, key, value) {
        data.set(key, value);
        return true;
      },
      has(target, prop) {
        return data.has(prop) || prop in target;
      },
      ownKeys(target) {
        const keys = [...Object.keys(target), ...data.keys()];
        return [...new Set(keys)];
      },

      /**
       * Lodash cloneDeep will use `Object.create(Object.getPrototypeOf(file))` which do not map to
       * the correct context. We need do a sub class to make skip fetch the context and still
       * instance of File. ref: https://github.com/ant-design/ant-design/issues/29646
       */
      getPrototypeOf() {
        class ProxyFile extends File {}

        const fileProtoKeys = Object.keys(File.prototype);

        [...fileProtoKeys, 'size', 'type'].forEach(key => {
          Object.defineProperty(ProxyFile.prototype, key, {
            // Get will never reach but we provide fallback here
            /* istanbul ignore next */
            get: () => getValue(key),
          });
        });

        return ProxyFile.prototype;
      },
      getOwnPropertyDescriptor(target, prop) {
        if (data.has(prop)) {
          return {
            value: data.get(prop),
            writable: true,
            enumerable: true,
            configurable: true,
          };
        }

        const descriptor = Object.getOwnPropertyDescriptor(target, prop);
        return descriptor;
      },
    });
  }

  return {
    ...file,
    ...filledProps,
  } as UploadFile;
}

export function replaceFileList(file: UploadFile<any>, fileList: UploadFile<any>[]) {
  const nextFileList = [...fileList];
  const fileIndex = nextFileList.findIndex(({ uid }: UploadFile) => uid === file.uid);
  if (fileIndex === -1) {
    nextFileList.push(file);
  } else {
    nextFileList[fileIndex] = file;
  }
  return nextFileList;
}

export function getFileItem(file: UploadFile, fileList: UploadFile[]) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name';
  return fileList.filter(item => item[matchKey] === file[matchKey])[0];
}

export function removeFileItem(file: UploadFile, fileList: UploadFile[]) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name';
  const removed = fileList.filter(item => item[matchKey] !== file[matchKey]);
  if (removed.length === fileList.length) {
    return null;
  }
  return removed;
}

// ==================== Default Image Preview ====================
const extname = (url: string = '') => {
  const temp = url.split('/');
  const filename = temp[temp.length - 1];
  const filenameWithoutSuffix = filename.split(/#|\?/)[0];
  return (/\.[^./\\]*$/.exec(filenameWithoutSuffix) || [''])[0];
};

const isImageFileType = (type: string): boolean => type.indexOf('image/') === 0;

export const isImageUrl = (file: UploadFile): boolean => {
  if (file.type && !file.thumbUrl) {
    return isImageFileType(file.type);
  }
  const url: string = (file.thumbUrl || file.url) as string;
  const extension = extname(url);
  if (
    /^data:image\//.test(url) ||
    /(webp|svg|png|gif|jpg|jpeg|jfif|bmp|dpg|ico)$/i.test(extension)
  ) {
    return true;
  }
  if (/^data:/.test(url)) {
    // other file types of base64
    return false;
  }
  if (extension) {
    // other file types which have extension
    return false;
  }
  return true;
};

const MEASURE_SIZE = 200;
export function previewImage(file: File | Blob): Promise<string> {
  return new Promise(resolve => {
    if (!file.type || !isImageFileType(file.type)) {
      resolve('');
      return;
    }

    const canvas = document.createElement('canvas');
    canvas.width = MEASURE_SIZE;
    canvas.height = MEASURE_SIZE;
    canvas.style.cssText = `position: fixed; left: 0; top: 0; width: ${MEASURE_SIZE}px; height: ${MEASURE_SIZE}px; z-index: 9999; display: none;`;
    document.body.appendChild(canvas);
    const ctx = canvas.getContext('2d');
    const img = new Image();
    img.onload = () => {
      const { width, height } = img;

      let drawWidth = MEASURE_SIZE;
      let drawHeight = MEASURE_SIZE;
      let offsetX = 0;
      let offsetY = 0;

      if (width > height) {
        drawHeight = height * (MEASURE_SIZE / width);
        offsetY = -(drawHeight - drawWidth) / 2;
      } else {
        drawWidth = width * (MEASURE_SIZE / height);
        offsetX = -(drawWidth - drawHeight) / 2;
      }

      ctx!.drawImage(img, offsetX, offsetY, drawWidth, drawHeight);
      const dataURL = canvas.toDataURL();
      document.body.removeChild(canvas);

      resolve(dataURL);
    };
    img.src = window.URL.createObjectURL(file);
  });
}
