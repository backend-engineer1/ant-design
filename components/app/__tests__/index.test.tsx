import React from 'react';
import App from '..';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { render } from '../../../tests/utils';

describe('App', () => {
  mountTest(App);
  rtlTest(App);

  it('single', () => {
    // Sub page
    const MyPage = () => {
      const { message } = App.useApp();
      React.useEffect(() => {
        message.success('Good!');
      }, [message]);

      return <div>Hello World</div>;
    };

    // Entry component
    const MyApp = () => (
      <App>
        <MyPage />
      </App>
    );

    const { getByText, container } = render(<MyApp />);
    expect(getByText('Hello World')).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
