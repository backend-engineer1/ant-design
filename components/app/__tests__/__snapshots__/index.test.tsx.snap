// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > rtl render > component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-app"
/>
`;

exports[`App > single 1`] = `
<div
  class="ant-app"
>
  <div>
    Hello World
  </div>
</div>
`;
